node/impl/full/eth_trace.go

```diff
@@ -38,212 +41,438 @@ func decodePayload(payload []byte, codec uint64) (ethtypes.EthBytes, error) {
-			// 3) Treat this as a delegate call to actor A.
-			if parent.LastByteCode != nil && trace.Action.From == trace.Action.To &&
-				trace.Action.FilecoinMethod == builtin.MethodsEVM.InvokeContractDelegate {
-				log.Debugf("COND7 found delegate call, height: %d", height)
-				prev := parent.LastByteCode
-				if prev.Action.From == trace.Action.From && prev.Action.FilecoinMethod == builtin.MethodsEVM.GetBytecode && prev.Parent == trace.Parent {
-					trace.SetCallType("delegatecall")
-					trace.Action.To = prev.Action.To
-
-					var dp evm.DelegateCallParams
-					err := dp.UnmarshalCBOR(bytes.NewReader(et.Msg.Params))
-					if err != nil {
-						return xerrors.Errorf("failed UnmarshalCBOR: %w", err)
-					}
-					trace.Action.Input = dp.Input
+			// Contract creation has no "to".
+			trace.Action.To = nil
 
-					trace.Result.Output, err = decodePayload(et.MsgRct.Return, et.MsgRct.ReturnCodec)
+			// Decode inputs & determine create type.
+			method, initcode, err := decodeCreate(&et.Msg)
+			if err != nil {
+				return nil, nil, xerrors.Errorf("EAM called with invalid params, but it still tried to construct the contract: %w", err)
+			}
+			trace.Action.Input = initcode
+			trace.SetCallType(method)
+
+			// Handle the output.
+			if et.MsgRct.ExitCode.IsError() {
+				if subTrace.MsgRct.ExitCode.IsError() {
+					// If we managed to call the constructor, parse/return its
+					// revert message.
+					output, err := decodePayload(subTrace.MsgRct.Return, subTrace.MsgRct.ReturnCodec)
 					if err != nil {
-						return xerrors.Errorf("failed decodePayload: %w", err)
+						log.Debugf("EVM actor returned indecipherable create error: %w", err)
+						output = encodeFilecoinReturnAsABI(
+							subTrace.MsgRct.ExitCode,
+							subTrace.MsgRct.ReturnCodec,
+							subTrace.MsgRct.Return,
+						)
 					}
+					trace.Result.Output = output
+				} else {
+					// Otherwise, if we failed before that, we have no revert
+					// message.
+					trace.Result.Output = nil
 				}
 			} else {
-				// Handle EVM call special casing
-				//
-				// Any outbound call from an EVM actor on methods 1-1023 are side-effects from EVM instructions
-				// and should be dropped from the trace.
-				if et.Msg.Method > 0 &&
-					et.Msg.Method <= 1023 {
-					log.Debugf("Infof found outbound call from an EVM actor on method 1-1023 method:%d, code:%s, height:%d", et.Msg.Method, parent.Action.FilecoinCodeCid.String(), height)
-
-					if et.Msg.Method == builtin.MethodsEVM.GetBytecode {
-						// save the last bytecode trace to handle delegate calls
-						parent.LastByteCode = trace
-					}
-
-					return nil
-				}
+				// We're _supposed_ to include the contracts bytecode here, but we
+				// can't do that reliably (e.g., if some part of the trace reverts).
+				// So we don't try and include a sentinel "impossible bytecode"
+				// value (the value specified by EIP-3541).
+				trace.Result.Output = []byte{0xFE}
```

Comment:
So, I'm supposed to return the contract's bytecode here. Unfortunately, I can't do that reliably because reverts (the bytecode may not actually exist in the state after the block is executed).

Options include:

1. Adding this information to the trace (that's how we've fixed all of our other "revert" issues). But, I _really_ don't want to do that.
2. Keep this as 0xFE.
3. Omit `Output` entirely?
4. Return `null`?

Reply:
You know, now that i think of it, no. Because:

1. We only flush the state at the end of the block, not at the end of every transaction.
2. A contract can be repeatedly created/destroyed (with the same init code but different _bytecode_), once per transaction.

That is...

TxA:

1. ContractA: calls Create2 with INITCODE.
2. ContractT (new): INITCODE loads the bytecode from the caller, and returns it.
3. ContractA: does some stuff, then calls back into ContractT.
4. ContractT: Calls SELFDESTRUCT.

TxB:

1. ContractB: calls Create2 with INITCODE (same INITCODE as above).
2. ContractT (re-created): INITCODE loads the bytecode from the caller (ContractB this time) and returns it. This is now _different_ bytecode.
3. ContractB: does some stuff, then calls back into ContractT.
4. ContractT: Calls SELFDESTRUCT.

By the end of the block, even with no reverted transactions, the ContractT could have changed its bytecode any number of times.

Reply:
This is fixable, it just requires even more changes to the FVM:

1. Record an actors state-object when it returns (instead of only when we call it).
2. Add a flag to keep all the intermediate state, and use it when tracing. But that could produce a _lot_ of garbage and I'd only be comfortable doing it if we passed in a special _separate_ datastore per-trace.

The upside is that this would give us nearly complete information on intermediate states in the system. So it's not a _bad_ idea, just something that would likely have to happen in a followup PR.